tests/gov_test.go

```diff
@@ -0,0 +1,69 @@
+package tests
+
+import (
+	"math/big"
+	"os"
+	"testing"
+
+	"github.com/kaiachain/kaia/blockchain"
+	"github.com/kaiachain/kaia/blockchain/types"
+	"github.com/kaiachain/kaia/governance"
+	"github.com/kaiachain/kaia/kaiax/gov"
+	gov_impl "github.com/kaiachain/kaia/kaiax/gov/impl"
+	"github.com/kaiachain/kaia/log"
+	"github.com/kaiachain/kaia/params"
+	"github.com/kaiachain/kaia/storage/database"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+// TestMainnetGenesisGovernance is a regression test for genesis parameters.
+// TODO-kaiax: replace mixedEngine with fixed genesis parameters values.
+func TestMainnetGenesisGovernance(t *testing.T) {
+	log.EnableLogForTest(log.LvlCrit, log.LvlError)
+	config := params.MainnetChainConfig.Copy()
+	genesis := blockchain.DefaultGenesisBlock()
+	genesis.Config = config
+	genesis.Governance = blockchain.SetGenesisGovernance(genesis)
+
+	fullNode, node, _, _, workspace := newBlockchain(t, config, genesis)
+	chain := node.BlockChain().(*blockchain.BlockChain)
+	defer os.RemoveAll(workspace)
+	defer fullNode.Stop()
+
+	// Create MixedEngine
+	dbm := database.NewMemoryDBManager()
+	mixed := governance.NewMixedEngine(config, dbm)
+
+	// Get params from MixedEngine
+	blockNum := uint64(0)
+	mixedParams, err := mixed.EffectiveParams(blockNum)
+	assert.NoError(t, err)
+
+	// Get params from GovModule
+	govModule := gov_impl.NewGovModule()
+	err = govModule.Init(&gov_impl.InitOpts{
+		Fallback:    map[gov.ParamName]interface{}{gov.GovernanceDeriveShaImpl: uint64(types.ImplDeriveShaConcat)},
```

Comment:
Okay. But are we going to permanently pass Fallback in node/cn/backend.go? How did MixedEngine got it correct at the first place?

Reply:
> are we going to permanently pass Fallback in node/cn/backend.go?

I think we have to..? Or we can set a default fallback in `gov.Init` in case of Mainnet/Kairos chain ID.

`params.ChainConfig` is `types.ImplDeriveShaConcat`, and the fallback chain was ContractGov -> HeaderGov -> ChainConfig -> Default values.

Reply:
I found that `governance.deriveshaimpl` is in genesis.json's `config` (via `mainnetGenesisJson` or `ken dumpgenesis`), but not in `header.governanceData` (via `kaia.getBlock(0).governanceData`). Apparently because [SetGenesisGovernance()](https://github.com/kaiachain/kaia/blob/e9fc051923c2c31877f7680a6fefeda9476f60f0/blockchain/genesis.go#L464) irreversibly missed it.

To generalize, [Config != Default && Genesis = Undefined] issue affects any other parameters that are missing from SetGenesisGovernance. For instance a private net with MagmaBlock=0 and LowerboundBaseFee=0gkei will suffer the same issue where Config says 0gkei, Default says 25gkei and Genesis does not specify it.

Don't we need to restore the fallback-to-Config logic?

Reply:
You may utilize the ReadChainConfig(genesisBlockHash) for clarity, instead of `*ChainConfig` variable.

Reply:
> a private net with MagmaBlock=0 and LowerboundBaseFee=0gkei will suffer the same issue where Config says 0gkei

This must've been fixed by [this PR](https://github.com/klaytn/klaytn/pull/1711); all parameters that are effective since genesis must be in the genesis header.Governance.

Reply:
Without ChainConfig fallback, it can be problematic for a private net where Magma > 0 and `genesis.json` set LowerBoundBaseFee to zero, in which case it'll fallback to 25 gkei.

Reply:
But you can't fix Genesis block header retrospectively. Mainnet's block 0 governanceData is 
```
kaia.getBlock(0).governanceData
0xb901c17b22676f7665726e616e63652e676f7665726e616e63656d6f6465223a2273696e676c65222c22676f7665726e616e63652e676f7665726e696e676e6f6465223a22307835326434316361373261663631356131616333333031623061393365666132323265636337353431222c22676f7665726e616e63652e756e69747072696365223a32353030303030303030302c22697374616e62756c2e636f6d6d697474656573697a65223a32322c22697374616e62756c2e65706f6368223a3630343830302c22697374616e62756c2e706f6c696379223a322c227265776172642e64656665727265647478666565223a747275652c227265776172642e6d696e696d756d7374616b65223a2235303030303030222c227265776172642e6d696e74696e67616d6f756e74223a2239363030303030303030303030303030303030222c227265776172642e70726f706f736572757064617465696e74657276616c223a333630302c227265776172642e726174696f223a2233342f35342f3132222c227265776172642e7374616b696e67757064617465696e74657276616c223a38363430302c227265776172642e75736567696e69636f656666223a747275657d
```
and it decodes to 
```json
{"governance.governancemode":"single","governance.governingnode":"0x52d41ca72af615a1ac3301b0a93efa222ecc7541","governance.unitprice":25000000000,"istanbul.committeesize":22,"istanbul.epoch":604800,"istanbul.policy":2,"reward.deferredtxfee":true,"reward.minimumstake":"5000000","reward.mintingamount":"9600000000000000000","reward.proposerupdateinterval":3600,"reward.ratio":"34/54/12","reward.stakingupdateinterval":86400,"reward.useginicoeff":true}
```

If GovModule refers to header.Governance, it will fall back to DefaultDeriveSha=0. And same for other parameters like lowerboundbasefee, upperboundbasefee, etc. PR 1711 does fix [parent block's BaseFee is out of bounds] issue but does not rewrite Genesis block.